larpandoracontent/LArTwoDReco/LArCosmicRay/TrackInEMShowerAlgorithm.cc

```diff
@@ -0,0 +1,877 @@
+    microFitResult.GetLocalPosition(splitPosition, rL, rT);
+
+    const TwoDSlidingFitResult macroFitResult(macroSlidingFitResultMap.at(pCluster));
+    CartesianVector averageDirection(0.f, 0.f, 0.f);
+    macroFitResult.GetGlobalDirection(macroFitResult.GetLayerFitResultMap().begin()->second.GetGradient(), averageDirection);
+
+    const bool isVertical(std::fabs(averageDirection.GetX()) < std::numeric_limits<float>::epsilon());
+    const float clusterGradient(isVertical ? 0.f : averageDirection.GetZ() / averageDirection.GetX());
+    const float clusterIntercept(isVertical ? splitPosition.GetX() : splitPosition.GetZ() - (clusterGradient * splitPosition.GetX()));
+
+    // Fragmentation initialisation
+    std::string originalListName, fragmentListName;
+    const ClusterList originalClusterList(1, pCluster);
+    PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::InitializeFragmentation(*this, originalClusterList, originalListName, fragmentListName));
+
+    const Cluster *pMainTrackCluster(nullptr), *pAboveCluster(nullptr), *pBelowCluster(nullptr);
+
+    const OrderedCaloHitList &orderedCaloHitList(pCluster->GetOrderedCaloHitList());
+    for (const OrderedCaloHitList::value_type &mapEntry : orderedCaloHitList)
+    {
+        for (const CaloHit *const pCaloHit : *mapEntry.second)
+        {
+            const CartesianVector &hitPosition(pCaloHit->GetPositionVector());
+            float thisL(0.f), thisT(0.f);
+
+            microFitResult.GetLocalPosition(pCaloHit->GetPositionVector(), thisL, thisT);
+
+            const bool isAnExtrapolatedHit(std::find(extrapolatedCaloHitIter->second.begin(), extrapolatedCaloHitIter->second.end(), pCaloHit) != extrapolatedCaloHitIter->second.end());
+            const bool isAbove(((clusterGradient * hitPosition.GetX()) + clusterIntercept) < (isVertical ? hitPosition.GetX() : hitPosition.GetZ()));
+            const bool isToRemove(!isAnExtrapolatedHit && (((thisL > rL) && isUpstream) || ((thisL < rL) && !isUpstream)));
+
+            const Cluster *&pClusterToModify(isToRemove ? (isAbove ? pAboveCluster : pBelowCluster) : pMainTrackCluster);
+
+            if (pClusterToModify)
+            {
+                PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::AddToCluster(*this, pClusterToModify, pCaloHit));
+            }
+            else
+            {
+                PandoraContentApi::Cluster::Parameters parameters;
+                parameters.m_caloHitList.push_back(pCaloHit);
+                PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::Cluster::Create(*this, parameters, pClusterToModify));
+
+                if (pClusterToModify != pMainTrackCluster)
+                    remnantClusterList.push_back(pClusterToModify);
+            }
+        }
+    }
+
+    this->UpdateForClusterDeletion(pCluster, clusterVector, microSlidingFitResultMap, macroSlidingFitResultMap);
```

Comment:
(This does seem like the right place for this update, but one more review of this and we should be happy 😄)